src/DotNetty.Handlers/Tls/SniHandler.cs

```diff
@@ -0,0 +1,260 @@
+                                    offset += 2;
+                                     int extensionsLimit = offset + extensionsLength;
+
+                                    if (extensionsLimit > endOffset)
+                                    {
+                                        // Extensions should never exceed the record boundary.
+                                        continueLoop = false;
+                                        break;
+                                    }
+
+                                    for (;;)
+                                    {
+                                        if (extensionsLimit - offset < 4)
+                                        {
+                                            continueLoop = false;
+                                            break;
+                                        }
+
+                                        int extensionType = input.GetUnsignedShort(offset);
+                                        offset += 2;
+
+                                        int extensionLength = input.GetUnsignedShort(offset);
+                                        offset += 2;
+
+                                        if (extensionsLimit - offset < extensionLength)
+                                        {
+                                            continueLoop = false;
+                                            break;
+                                        }
+
+                                        // SNI
+                                        // See https://tools.ietf.org/html/rfc6066#page-6
+                                        if (extensionType == 0)
+                                        {
+                                            offset += 2;
+                                            if (extensionsLimit - offset < 3)
+                                            {
+                                                continueLoop = false;
+                                                break;
+                                            }
+
+                                            int serverNameType = input.GetByte(offset);
+                                            offset++;
+
+                                            if (serverNameType == 0)
+                                            {
+                                                int serverNameLength = input.GetUnsignedShort(offset);
+                                                offset += 2;
+
+                                                if (serverNameLength <= 0 || extensionsLimit - offset < serverNameLength)
+                                                {
+                                                    continueLoop = false;
+                                                    break;
+                                                }
+
+                                                string hostname = input.ToString(offset, serverNameLength, Encoding.ASCII);
```

Comment:
use Encoding.UTF8.GetString() to parse the string out of bytes (per RFC for TLS Extensions), then use IdnMapping.ToAscii()